src/main/java/com/cburch/logisim/std/ttl/Ttl74192.java

```diff
@@ -0,0 +1,232 @@
+    }
+  }
+
+  @Override
+  public void paintInternal(InstancePainter painter, int x, int y, int height, boolean up) {
+    final var gfx = (Graphics2D) painter.getGraphics();
+    super.paintBase(painter, false, false);
+    Drawgates.paintPortNames(
+        painter,
+        x,
+        y,
+        height,
+        new String[] {
+          "B", "QB", "QA", "CntD", "CntU", "QC", "QD", "D", "C", "LOAD", "CAR", "BOR", "CLR", "A"
+        });
+    UpDownCounterData data = (UpDownCounterData) painter.getData();
+    drawState(gfx, x, y, height, data);
+  }
+
+  private void drawState(Graphics2D gfx, int x, int y, int height, UpDownCounterData state) {
+    if (state == null) return;
+    var value = state.getValue();
+    for (var i = 0; i < 4; i++) {
+      final var bitValue = value.get(3 - i);
+      gfx.setColor(bitValue.getColor());
+      gfx.fillOval(x + 52 + i * 10, y + height / 2 - 4, 8, 8);
+      gfx.setColor(Color.WHITE);
+      GraphicsUtil.drawCenteredText(gfx, bitValue == Value.TRUE ? "1" : "0", x + 56 + i * 10, y + height / 2);
+    }
+    gfx.setColor(Color.BLACK);
+  }
+
+  public static void updateState(InstanceState state, Value value, Value carry, Value borrow, Value down, Value up) {
+    final var data = getStateData(state);
+
+    data.setAll(value, carry, borrow, down, up);
+    final var vA = data.getValue().get(0);
+    final var vB = data.getValue().get(1);
+    final var vC = data.getValue().get(2);
+    final var vD = data.getValue().get(3);
+    final var vCar = data.getCarry();
+    final var vBor = data.getBorrow();
+
+    state.setPort(PORT_INDEX_QA, vA, 4);
+    state.setPort(PORT_INDEX_QB, vB, 4);
+    state.setPort(PORT_INDEX_QC, vC, 4);
+    state.setPort(PORT_INDEX_QD, vD, 4);
+    state.setPort(PORT_INDEX_CARRY, vCar, 4);
+    state.setPort(PORT_INDEX_BORROW, vBor, 4);
+  }
+
+  public static UpDownCounterData getStateData(InstanceState state) {
+    var data = (UpDownCounterData) state.getData();
+    if (data == null) {
+      data = new UpDownCounterData();
+      state.setData(data);
+    }
+    return data;
+  }
+
+  @Override
+  public void propagateTtl(InstanceState state) {
+    final var data = getStateData(state);
+
+    var carry = Value.TRUE;
+    var borrow = Value.TRUE;
+    var counter = data.getValue().toLongValue();
+
+    final var downPrev = data.getDownPrev();
+    final var upPrev = data.getUpPrev();
+    final var downCur = state.getPortValue(PORT_INDEX_DOWN);
+    final var upCur = state.getPortValue(PORT_INDEX_UP);
+    final var downFalling = downPrev == Value.TRUE && downCur == Value.FALSE;
+    final var downRising = downPrev == Value.FALSE && downCur == Value.TRUE;
+    final var upFalling = upPrev == Value.TRUE && upCur == Value.FALSE;
+    final var upRising = upPrev == Value.FALSE && upCur == Value.TRUE;
+    final var downUnchangedHigh = downPrev == Value.TRUE && downCur == Value.TRUE;
+    final var upUnchangedHigh = upPrev == Value.TRUE && upCur == Value.TRUE;
```

Comment:
I'm not really happy with this solution, but currently can't think of a different way.

Reply:
Sometimes it is not worth spending too much time polishing non important things. You may want to add `FIXME` note to that code block if you need, describing your thoughts and why it should be refactored and probably leave for now. That's not a bad code.